src/main/java/io/debezium/connector/jdbc/JdbcChangeEventSink.java

```diff
@@ -40,71 +42,181 @@ public class JdbcChangeEventSink implements ChangeEventSink {
+
+            validate(record);
+
+            Optional<TableId> optionalTableId = getTableId(record);
+            if (optionalTableId.isEmpty()) {
+
+                LOGGER.warn("Ignored to write record from topic '{}' partition '{}' offset '{}'. No resolvable table name", record.topic(), record.kafkaPartition(),
+                        record.kafkaOffset());
+                continue;
+            }
+
+            SinkRecordDescriptor sinkRecordDescriptor = buildRecordSinkDescriptor(record);
+
+            final TableId tableId = optionalTableId.get();
+
+            if (sinkRecordDescriptor.isTombstone()) {
+                LOGGER.debug("Skipping tombstone record {}", sinkRecordDescriptor);
+                continue;
+            }
+
+            if (sinkRecordDescriptor.isTruncate()) {
+
+                if (!config.isTruncateEnabled()) {
+                    LOGGER.debug("Truncates are not enabled, skipping truncate for topic '{}'", sinkRecordDescriptor.getTopicName());
+                    continue;
+                }
+
+                // Here we want to flush the buffer to let truncate having effect on the buffered events.
+                flushBuffers(updateBufferByTable);
+
+                flushBuffers(deleteBufferByTable);
+
+                try {
+                    final TableDescriptor table = checkAndApplyTableChangesIfNeeded(tableId, sinkRecordDescriptor);
+                    writeTruncate(dialect.getTruncateStatement(table));
+                }
+                catch (SQLException e) {
+                    throw new ConnectException("Failed to process a sink record", e);
+                }
+            }
+
+            if (sinkRecordDescriptor.isDelete()) {
+
+                if (!config.isDeleteEnabled()) {
+                    LOGGER.debug("Deletes are not enabled, skipping delete for topic '{}'", sinkRecordDescriptor.getTopicName());
+                    continue;
+                }
+
+                RecordBuffer tableIdBuffer = deleteBufferByTable.computeIfAbsent(tableId, k -> new RecordBuffer(config));
+                List<SinkRecordDescriptor> toFlush = tableIdBuffer.add(sinkRecordDescriptor);
+
+                flushBuffer(tableId, toFlush);
+            }
+            else {
+
+                if (deleteBufferByTable.get(tableId) != null && !deleteBufferByTable.get(tableId).isEmpty()) {
+                    // When an insert arrives, delete buffer must be flushed to avoid losing an insert for the same record after its deletion.
+                    // this because at the end we will always flush inserts before deletes.
+
+                    flushBuffer(tableId, deleteBufferByTable.get(tableId).flush());
+                }
+
+                RecordBuffer tableIdBuffer = updateBufferByTable.computeIfAbsent(tableId, k -> new RecordBuffer(config));
+
+                List<SinkRecordDescriptor> toFlush = tableIdBuffer.add(sinkRecordDescriptor);
+                flushBuffer(tableId, toFlush);
```

Comment:
So, in this case, we have received an Insert or an Update event, so can you explain why we need to flush here when adding that sink record to the buffer?  The delete use case makes sense, but this one I'm not exactly clear on.

Reply:
Given that, a the end of a `put` batch,  at line [143](https://github.com/debezium/debezium-connector-jdbc/pull/43/files#diff-e6057968f3afc7c4708d184385030eeae203de011f7336c40fce6a5d5eb5c7a3R143)  and [145](https://github.com/debezium/debezium-connector-jdbc/pull/43/files#diff-e6057968f3afc7c4708d184385030eeae203de011f7336c40fce6a5d5eb5c7a3R145)  the `update/insert` buffer will be flushed before the `delete` one and suppose that you have 

1. Insert of record with id X 
2. Delete of record with id X
3. Insert of record with id X.

The end result on the database should be the record `#3`. 

Since the buffering of `insert/update` records are different from the `deletes` one and  then at the end the `insert/update` buffer will be flushed first, if we don't flush the delete when it's arrives we will loose the order.  As result, This cause the loose of the record with id X.